okhttp/src/main/java/okhttp3/NewWebSocket.java

```diff
@@ -0,0 +1,150 @@
+  /** Returns the original request that initiated this websocket. */
+  Request request();
+
+  /**
+   * Returns the size in bytes of all messages enqueued to be transmitted to the server. This
+   * doesn't include framing overhead. It also doesn't include any bytes buffered by the operating
+   * system or network intermediaries. This method returns 0 if no messages are waiting
+   * in the queue. If may return a nonzero value after the websocket has been canceled; this
+   * indicates that enqueued messages were not transmitted.
+   */
+  long queueSize();
+
+  /**
+   * Attempts to enqueue {@code text} to be UTF-8 encoded and sent as a the data of a text (type
+   * {@code 0x1}) message.
+   *
+   * <p>This method returns true if the message was enqueued. Messages that would overflow the
+   * outgoing message buffer will be rejected and trigger a {@link #close graceful shutdown} of
+   * this websocket. This method returns false in that case, and in any other case where this
+   * websocket is closing, closed, or canceled.
+   *
+   * <p>This method returns immediately.
+   */
+  boolean send(String text);
+
+  /**
+   * Attempts to enqueue {@code bytes} to be sent as a the data of a binary (type {@code 0x2})
+   * message.
+   *
+   * <p>This method returns true if the message was enqueued. Messages that would overflow the
+   * outgoing message buffer will be rejected and trigger a {@link #close graceful shutdown} of
+   * this websocket. This method returns false in that case, and in any other case where this
+   * websocket is closing, closed, or canceled.
+   *
+   * <p>This method returns immediately.
+   */
+  boolean send(ByteString bytes);
+
+  /**
+   * Attempts to initiate a graceful shutdown of this websocket. Any already-enqueued messages will
+   * be transmitted before the close message is sent but subsequent calls to {@link #send} will
+   * return false and their messages will not be enqueued.
+   *
+   * <p>This returns true if a graceful shutdown was initiated by this call. It returns false and if
+   * a graceful shutdown was already underway or if the websocket is already closed or canceled.
+   *
+   * @param code Status code as defined by <a
+   * href="http://tools.ietf.org/html/rfc6455#section-7.4">Section 7.4 of RFC 6455</a> or {@code 0}.
+   * @param reason Reason for shutting down or {@code null}.
+   */
+  boolean close(int code, String reason);
+
+  /**
+   * Immediately and violently release resources held by this websocket, discarding any enqueued
+   * messages. This does nothing if the websocket has already been closed or canceled.
+   */
+  void cancel();
+
+  interface Factory {
+    NewWebSocket newWebSocket(Request request, Listener listener);
+  }
+
+  abstract class Listener {
+    /**
+     * Invoked when a websocket has been accepted by the remote peer and may begin transmitting
+     * messages.
+     */
+    public void onOpen(NewWebSocket websocket, Response response) {
```

Comment:
I don't think this needs to return the `NewWebSocket` anymore since you get it with the `Factory`.